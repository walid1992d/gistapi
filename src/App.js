
import { useState, useMemo, useEffect } from 'react';
import styled from 'styled-components'
import GistList from './components/GistList';
import Header from "./components/Header";
import { GistContext } from './contexts/gistContext';
import { UsernameContext } from './contexts/usernameContext';
import GlobalStyles from "./GlobalStyle";
import { getGistForUser, getPublicGists } from './services/gistService';

const App = () => {
  const [usernameState, setUserName] = useState();
  const [gistListState, setGistList] = useState([]);
  const [screenMessage, setScreenMessage] = useState('');
  const usernameProviderValue = useMemo(() => ({ usernameState, setUserName}), [usernameState, setUserName]);
  const gistListProvider = useMemo(() => ({gistListState, setGistList}), [gistListState, setGistList]);
  useEffect( ()=> {
    const fetchData = async() => {
      try {
      let response;
      if(!usernameState) {
        response = await getPublicGists();
      } else {
        response = await getGistForUser(usernameState);
        
      }

      const {status, data, message} = response;
      if(status === 200) {
        if(data.length === 0) {
          setScreenMessage("No Results Found");

        } else {
          setGistList(data);
          setScreenMessage("");
          
        }

      }  else {
        setScreenMessage("Something Went Wrong :(");
      }

     } catch(e) {
      setScreenMessage("Something Went Wrong :(");

     }
    }

    fetchData();
    
  },[usernameState])
  return (

    <Wrapper className="App" data-testid="app">
      <UsernameContext.Provider value={usernameProviderValue} >
      <GistContext.Provider value={gistListProvider}>
      <Header />
      {
        screenMessage ? <ScreenMessageDiv>{screenMessage}</ScreenMessageDiv> :  <GistListWrapper><GistList>  </GistList></GistListWrapper>

      }
     
    
      </GistContext.Provider>

      </UsernameContext.Provider>
      <GlobalStyles />
    </Wrapper>
  );
}

const Wrapper = styled.div`
  font-size: 14px;
  line-height: 1.5;
`;

const ScreenMessageDiv = styled.div `
  display: flex;
  width: 100%;
  justify-content: center;
  padding: 2vw;
  font-size: 2vw;
`

const GistListWrapper = styled.div `
  width: 40%;
  min-width: 500px;
  margin: auto;

  @media (max-width: 500px) {
   width: 90%;
   min-width: unset;
  }
`;

export default App;
