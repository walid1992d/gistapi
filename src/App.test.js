import { act, findByTestId, render } from '@testing-library/react';
import App from './App';
import * as gistService from './services/gistService';
jest.mock('./components/Header', () => {
  return ()=> <div id="header"></div>
});
jest.mock('./components/GistList', () => {
  return ()=> <div id="list"></div>
});
describe('App', () => {
  it('should render correctly if fetching gists succeeds', async()=> {
    jest.spyOn(gistService,'getPublicGists').mockImplementation(() => {
      return {
        status: 200,
        data: [{}],
      }

  })
  let container
  await act(()=> {
     container = render(<App />).container;

  })
      expect(container).toMatchSnapshot();
  
  

  });
  it('should display error if fetching gists fails', async ()=> {
    jest.spyOn(gistService,'getPublicGists').mockImplementation(() => {
      return {
        status: 400,
        data: [],
      }
    });

      let container
      await act(()=> {
         container = render(<App />).container;
    
      });
      expect(container).toMatchSnapshot();

   
        

  });
  it('should display no results message if gits result is empty', async ()=> {
    jest.spyOn(gistService,'getPublicGists').mockImplementation(() => {
      return {
        status: 200,
        data: [],
      }
    });

      let container
      await act(()=> {
         container = render(<App />).container;
    
      });
      expect(container).toMatchSnapshot();

   
        

  });

});
