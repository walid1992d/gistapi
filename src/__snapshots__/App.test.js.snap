// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should display error if fetching gists fails 1`] = `
<div>
  <div
    class="sc-bdnxRM SuNwZ App"
    data-testid="app"
  >
    <div
      id="header"
    />
    <div
      class="sc-gtsrHT hyXrb"
    >
      Something Went Wrong :(
    </div>
  </div>
</div>
`;

exports[`App should display no results message if gits result is empty 1`] = `
<div>
  <div
    class="sc-bdnxRM SuNwZ App"
    data-testid="app"
  >
    <div
      id="header"
    />
    <div
      class="sc-gtsrHT hyXrb"
    >
      No Results Found
    </div>
  </div>
</div>
`;

exports[`App should render correctly if fetching gists succeeds 1`] = `
<div>
  <div
    class="sc-bdnxRM SuNwZ App"
    data-testid="app"
  >
    <div
      id="header"
    />
    <div
      class="sc-dlnjwi bnvrgC"
    >
      <div
        id="list"
      />
    </div>
  </div>
</div>
`;
